cng/dsa.go

```diff
@@ -0,0 +1,473 @@
+		consumeBigInt(hdr.SeedLength)
+		params = DSAParameters{
+			Q: consumeBigInt(hdr.GroupSize),
+			P: consumeBigInt(hdr.KeySize),
+			G: consumeBigInt(hdr.KeySize),
+		}
+		Y = consumeBigInt(hdr.KeySize)
+		if private {
+			X = consumeBigInt(hdr.GroupSize)
+		}
+	}
+	return params, X, Y, nil
+}
+
+// setDSAParameter sets the DSA parameters for the given key.
+func setDSAParameter(hkey bcrypt.KEY_HANDLE, params DSAParameters) error {
+	keySize := params.keySize()
+	groupSize := params.groupSize()
+	var blob []byte
+	if keySize*8 <= 1024 {
+		blob = make([]byte, sizeOfDSAParamsHeader+keySize*2)
+		hdr := bcrypt.DSA_PARAMETER_HEADER{
+			Length:  uint32(len(blob)),
+			Magic:   bcrypt.DSA_PARAMETERS_MAGIC,
+			KeySize: keySize,
+			Count:   dsaCountNil,
+		}
+		copy(hdr.Seed[:], dsaSeedNil[:])
+		copy(hdr.Q[:], params.Q[:])
+		copy(blob, (*(*[sizeOfDSAParamsHeader]byte)(unsafe.Pointer(&hdr)))[:])
+		data := blob[sizeOfDSAParamsHeader:]
+		if err := encodeBigInt(data, []sizedBigInt{
+			{params.P, keySize},
+			{params.G, keySize},
+		}); err != nil {
+			return err
+		}
+	} else {
+		blob = make([]byte, sizeOfDSAParamsV2Header+2*keySize+2*groupSize)
+		hashAlg := hashAlgFromGroup(int(groupSize))
+		hdr := bcrypt.DSA_PARAMETER_HEADER_V2{
+			Length:          uint32(len(blob)),
+			Magic:           bcrypt.DSA_PARAMETERS_MAGIC_V2,
+			KeySize:         keySize,
+			GroupSize:       groupSize,
+			HashAlgorithm:   hashAlg,
+			StandardVersion: bcrypt.DSA_FIPS186_3,
+			SeedLength:      groupSize, // crypto/dsa doesn't use the seed, but CNG expects it to be groupSize.
+			Count:           dsaCountNil,
+		}
+		copy(blob, (*(*[sizeOfDSAParamsV2Header]byte)(unsafe.Pointer(&hdr)))[:])
+		data := blob[sizeOfDSAParamsV2Header:]
+		if err := encodeBigInt(data, []sizedBigInt{
+			{dsaSeedNil[:], groupSize},
+			{params.Q, groupSize},
+			{params.P, keySize},
+			{params.G, keySize},
+		}); err != nil {
+			return err
+		}
+
+	}
+	return bcrypt.SetProperty(bcrypt.HANDLE(hkey), utf16PtrFromString(bcrypt.DSA_PARAMETERS), blob, 0)
+}
+
+func dsaAdjustHashSize(hkey bcrypt.KEY_HANDLE, hashed []byte, buf []byte) ([]byte, error) {
+	// Windows CNG requires that the hash output and Q match sizes, but we can better
+	// interoperate with other FIPS 186-3 implementations if we perform truncation
+	// here, before sending it to CNG.
+	//
+	// If, on the other hand, Q is too big, we need to left-pad the hash with zeroes
+	// (since it gets treated as a big-endian number).
+	params, _, _, err := decodeDSAKey(hkey, false)
+	if err != nil {
+		return nil, err
+	}
+	groupSize := int(params.groupSize())
+	if groupSize > len(buf) {
+		panic("output buffer too small")
+	}
+	if groupSize == len(hashed) {
+		return hashed, nil
+	}
+	if groupSize < len(hashed) {
+		return hashed[:groupSize], nil
+	}
+	zeroByteCount := groupSize - len(hashed)
+	for i := 0; i < zeroByteCount; i++ {
+		buf[i] = 0
+	}
+	copy(buf[zeroByteCount:], hashed)
+	return buf[:groupSize], nil
```

Comment:
This seems a *little* similar to `encodeBigInt`... I don't know if it's feasible/worthwhile to actually share any logic or define this in terms of `sizedBigInt` though.

Reply:
Good catch. Zeroing the buffer is not necessary, as it is already empty, so we can reuse `dsaAdjustHashSize`.